### ISSUE TITLE
weird decorators not returning a callable are acceptable?

### ISSUE THREAD

**Issue by wbolster** [python/mypy#9752]
this rather weird-looking code that i [wrote as a joke](https://twitter.com/wbolster/status/1331320901044989954):

``` python
>>> def assign(v): return lambda f: v                                                                               
>>> @assign(1) 
... def a(): pass                                                                                                   
>>> print(a)                                                                                                        
1
```

...abuses decorator syntax to assign a random value `v` (instead of a replacement function) to the name `a`.

i am surprised that a type annotated version of this code:

``` python
from typing import Callable


def assign(v: object) -> Callable[[Callable[..., object]], object]:
    return lambda f: v


@assign(1)
def a() -> None:
    pass


print(a)
```

...manages to sneak past `mypy`:

``` shell
$ python test.py   # it still ‘works’
1

$ mypy --version
mypy 0.790

$ mypy --check-untyped-defs --strict test.py 
Success: no issues found in 1 source file
```

...even though the result of `assign()`, when used as a decorator, does *not* return a callable.

for what it's worth, without `lambda` the result is the same:

``` python
def assign(v: object) -> Callable[[Callable[..., object]], object]:             
    def decorator(f: Callable[..., object]) -> object:                          
        return v                                                                
    return decorator                                                            
```

am i missing something here? are my annotations wrong? is mypy too tolerant about decorator signatures? are there valid reasons to allow decorators to return something that is *not* callable*?

(sorry for the weird report. it's a bit puzzling for me. 😅)

**Comment by hauntsaninja** [comment]
It's valid code, so why should mypy complain. mypy infers the correct type for `a` in your example, so if you try to call `a` you'll get a TypeError.
